include/cutlass/fp8_to_fp16.h

```diff
@@ -117,12 +121,133 @@ static inline ushort16 E4M3_to_FP16_chunk16(uchar16 xin) {
+    uint32_t fifth = (tmp1 & 0x000000FF) << 8;
+    uint32_t sixth = (tmp1 & 0x0000FF00) << 16;
+    uint32_t seventh = (tmp1 & 0x00FF0000) >> 8;
+    uint32_t eighth = (tmp1 & 0xFF000000);
+    xout[i] = first | second;
+    xout[i + 1] = third | fourth;
+    xout[i + 2] = fifth | sixth;
+    xout[i + 3] = seventh | eighth;
+  }
+}
 
-template<int N>
-static inline void E5M2_to_FP16(cutlass::Array<uint8_t, N> const &xin, cutlass::Array<uint16_t, N> &xout) {
-  // Adapted from https://github.com/pytorch/pytorch/blob/dfcfad2112933cc34247421ac0a4d3f19a1806c1/c10/util/Float8_e5m2.h#L30-L43
+template <int N>
+static inline void E5M2_to_FP16(cutlass::Array<uint8_t, N> const &xin,
+                                cutlass::Array<uint16_t, N> &xout) {
+  // Adapted from
+  // https://github.com/pytorch/pytorch/blob/dfcfad2112933cc34247421ac0a4d3f19a1806c1/c10/util/Float8_e5m2.h#L30-L43
+  // Using something as simple as the following code surprisingly
+  // leads to poor performance.
+  // CUTLASS_PRAGMA_UNROLL
+  // for (int i = 0; i < num_elements; i++) {
+  //   reinterpret_cast<uint16_t*>(pDst)[i] =
+  //   (static_cast<uint16_t>((pSrc[i]))) << 8;
+  // }
+  // The root-cause is unknown, but private memory use is seen in that case.
+  // We're using a workaround that doesn't use private memory.
   CUTLASS_PRAGMA_UNROLL
   for (int i = 0; i < N; i++) {
     xout[i] = (static_cast<uint16_t>(xin[i])) << 8;
   }
 }
+
+template <typename ElementA, class EngineIn, class EngineOut, class LayoutIn,
+          class LayoutOut, class... Ts>
+CUTLASS_DEVICE void
+convert_FP8_to_FP16(cute::Tensor<EngineIn, LayoutIn> const &in,
+                    cute::Tensor<EngineOut, LayoutOut> &out) {
+
+  static_assert(cute::is_rmem<EngineIn>::value,
+                "Input tensor for A conversion must come from registers");
+  static_assert(cute::is_rmem<EngineOut>::value,
+                "Output tensor for A conversion must come from registers");
+  static_assert(cute::cosize_v<LayoutIn> == cute::cosize_v<LayoutOut>);
+  static_assert(cute::size_v<LayoutIn> == cute::cosize_v<LayoutIn>);
+  static_assert(cute::size_v<LayoutOut> == cute::cosize_v<LayoutOut>);
+
+  using SrcType = typename EngineIn::value_type;
+  using DstType = typename EngineOut::value_type;
+
+  static_assert(std::is_same_v<SrcType, uint8_t>,
+                "Expected fp8 (E4M3) input as uint8_t");
+  static_assert(std::is_same_v<DstType, cute::half_t>,
+                "Expected fp16 output as half_t");
+
+  auto const &src = in(cute::_, cute::_, cute::_);
+  auto const &dst = out(cute::_, cute::_, cute::_);
+
+  SrcType const *pSrc = src.data();
+  DstType *pDst = dst.data();
+
+  constexpr int num_elements = decltype(size(src))::value;
+
+  // TODO(Codeplay): Move conversion to NumericArrayConverter
+  if constexpr (std::is_same_v<ElementA, cute::float_e5m2_t>) {
+    // May convert four FP8 elements at a time
+    constexpr bool use_faster_conversion = num_elements >= 8;
```

Comment:
```suggestion
    constexpr bool use_uint32_container = num_elements >= 8;
```

Reply:
Also, could you not use uint32 for all cases of at least 4 elements (so you do not need 8)?

Reply:
Loop unrolling is not working if I create a nested loop. :(
If it did, then I could've written code to select an unrolling factor at compile-time.

Since I was forced to unroll the nested loop manually (i.e. the nested loop isn't present in the code because it's been unrolled), I used `8` so that `8x16` subgroup tile conversion could also be covered down the line.

Reply:
> Also if you do, can you just manually unroll it?

It's already being done in the implementation in this PR - the inner loop has been manually unrolled to convert 8 elements in one iteration of the outer-loop, which is akin to having an inner loop with an unroll factor of 2 (since one nested loop iteration would convert 4 FP8 elements).

Since inner loop unrolling with preprocessor directive is not working, if I wanted to convert 4 elements or 16 elements per per work-item in one outer loop iteration, then I'd have to write separate code for them.

If inner loop unrolling worked, I could've decided the unroll factor at compile time.